tests/apitest/src/CoreServices/FSEventStream.cs

```diff
@@ -0,0 +1,177 @@
+﻿using System;
+using System.Threading;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Runtime.InteropServices;
+using CoreServices;
+using CoreFoundation;
+using ObjCRuntime;
+using Foundation;
+using NUnit.Framework;
+using Xamarin.Mac.Tests;
+using System.Threading.Tasks;
+
+namespace Xamarin.Mac.Tests {
+	[TestFixture]
+	public class FSEventStreamTests {
+
+		string dirPath;
+		DirectoryInfo dir;
+		bool watchedFileChanged;
+		bool exludedFileChanged;
+		ulong st_dev;
+		FSEventStream fsEventStream;
+		string [] pathsToWatchRelativeToDevice;
+
+		ulong GetDeviceId ()
+		{
+			var process = new Process () {
+				StartInfo = new ProcessStartInfo {
+					FileName = "/usr/bin/stat",
+					Arguments = $"-f '%d' {Path.Combine (dirPath, "TempFileToWatch.txt")}",
+					RedirectStandardOutput = true,
+					UseShellExecute = false,
+					CreateNoWindow = true,
+				}
+			};
+			process.Start ();
+			string result = process.StandardOutput.ReadToEnd ();
+			process.WaitForExit ();
+			return UInt64.Parse (result);
+		}
+
+		[TestFixtureSetUp]
+		public void Init ()
+		{
+			dirPath = Path.Combine (Path.GetTempPath (), "FSEventStreamTests");
+			st_dev = GetDeviceId ();
+
+			// The reason why we take from substring from index 1 is because in this particular API, the paths should be relative to
+			// the root of the device. So instead of "/Pictures/July", we need to provide "Pictures/July".
+			// Refer to https://developer.apple.com/documentation/coreservices/1447341-fseventstreamcreaterelativetodev?language=objc
+			pathsToWatchRelativeToDevice = new [] { dirPath.Substring (1) };
+			dir = Directory.CreateDirectory (dirPath);
+
+			using (FileStream fileToWatch = File.Create (Path.Combine (dirPath, "TempFileToWatch.txt")))
+			using (FileStream fileToExclude = File.Create (Path.Combine (dirPath, "TempFileToExclude.txt"))) { }
+		}
+
+		[TestFixtureTearDown]
+		public void Destroy ()
+		{
+			Directory.Delete ("FSEventStreamTests", true);
+		}
+
+		[SetUp]
+		public void CreateFSEventStream ()
+		{
+			watchedFileChanged = false;
+			exludedFileChanged = false;
+
+			fsEventStream = new FSEventStream (st_dev, pathsToWatchRelativeToDevice, FSEvent.SinceNowId, TimeSpan.FromSeconds (1), FSEventStreamCreateFlags.WatchRoot | FSEventStreamCreateFlags.FileEvents | FSEventStreamCreateFlags.NoDefer);
+			Assert.IsNotNull (fsEventStream, "Null");
+		}
+
+		[TearDown]
+		public void EndFSEventStream ()
+		{
+			fsEventStream.Stop ();
```

Comment:
Does this work if the test does not start the stream? We do not start it in the SetUp AFAIK. Could, if we ever add a test that does not start it, be an issue. Just double check.